h2o-core/src/main/java/hex/Model.java

```diff
@@ -1251,12 +1271,12 @@ else if (isWeights && computeMetrics) {
             vec = evec;
           }
         } else if(vec.isCategorical()) {
-          if (parms._categorical_encoding == Parameters.CategoricalEncodingScheme.LabelEncoder) {
+          try {
             Vec evec = vec.toNumericVec();
             toDelete.put(evec._key, "label encoded vec");
             vec = evec;
-          } else {
-            throw new IllegalArgumentException("Test/Validation dataset has categorical column '" + names[i] + "' which is real-valued in the training data");
+          } catch(Exception e) {
```

Comment:
suggesting to be more lenient here as the conversion above is actually half-backed imo.
It works only if the domains from the test frame are indexed exactly the same way as the training frame (which provides the domains through `_output.domains`).
Although, it can easily not be the case if test frame is a feature subset of training features for example.

Reply:
Not sure if we are on the same page - can you explain a bit more?

IMHO `toNumericVec` on a categorical column will try to interpret the domain values as integer numbers according to these rules:

   * If the first value in the domain of the src Vec is a stringified ints,
   * then it will use those ints. Otherwise, it will use the raw enumeration level mapping.

This behavior honestly doesn't make much sense to me right now and can be outright dangerous as it all depends just on the first value of the domain.

We get a test frame with a Vec that is categorical instead of numerical. IMO this has nothing to do with categorical encoding. The Model output remembers the original domain of the training frame (before label encoding...) and this means the original training Vec simply was numerical not categorical.

Am I missing something?

Reply:
Looks like LabelEncoder is broken: https://github.com/h2oai/h2o-3/pull/3527

Reply:
@michalkurka the test behaves as expected with my changes here.
It's not as much that `H2O forgets the original domain of LabelEncoded columns`: this is actually not the case, there are available in `_output._origDomains` whatever the encoding.
However, those origDomains are currently ignored in `adaptTestForTrain` for `LabelEncoder`, hence the reason I'm adding it to the `checkCategoricals` list, and also why I want to have it working with `EnumLimited` as well.

I have also good reasons to think that the absence of those 2 encodings there is an omission: those have been created (in FrameUtils) several months after `Binary`, `Eigen`... and given the complete decoupling between this logic and their definition in `FrameUtils`, it would not be surprising that author forgot to add them to the list here: I'll had a new method to the enum to avoid this in the future.

Reply:
regarding the call to `toNumericVec`: it is a fallback mechanism. It was called much more frequently before my changes (but only if encoding was specified to LabelEncoder).
Will check various use-cases to figure out if it shouldn't be simply removed.

Reply:
Perfect, that makes sense - thank you.